File: TouchControl/src/touchcontrol/touchables/TouchableGroup.java
/**
 * Class Description:
 * Apply the same function over many Touchable objects
 */

package touchcontrol.touchables;

import java.util.ArrayList;

import org.opencv.core.*;

public class TouchableGroup
{
	ArrayList<Touchable> components;

	//Constructors
	public TouchableGroup(Touchable...components)
	{
		this.components = new ArrayList<>();

		for (Touchable component : components)
			this.components.add(component);

		checkCollision();
	}

	//Setters
	public void setColor(Scalar color)
	{
		for (Touchable component : components)
			component.setColor(color);
	}

	public void addComponent(Touchable component)
	{
		components.add(component);

		checkCollision();
	}

	public void removeComponent(Touchable component)
	{
		components.remove(component);
	}

	//Getters
	public ArrayList<Touchable> getComponents()
	{
		return components;
	}

	//Class actions
	public void updateDetectionPoint(Mat filteredImage)
	{
		for (Touchable component : components)
			component.updateDetectionPoint(filteredImage);
	}

	public void drawOnto(Mat rawImage)
	{
		for (Touchable component : components)
			component.drawOnto(rawImage);
	}

	public void performAction()
	{
		for (Touchable component : components)
			component.performAction();
	}

	@Override
	public String toString()
	{
		String groupData = "Touchable objects in group (" + super.toString() + "): " + components.size() + " \n\n";

		for (int i = 0; i < components.size(); i++)
			groupData += "(" + i + ")" + components.get(i).toString() + "\n";

		return groupData;
	}

	private void checkCollision()
	{
		for (int i = 0; i < components.size(); i++)
			for (int j = i + 1; j < components.size(); j++)
			{
				//Get components dimensions
				Rect dim1 = components.get(i).getDimensions();
				Rect dim2 = components.get(j).getDimensions();

				//Check if two components overlap
				boolean overlaps = !(dim1.y + dim1.height <= dim2.y
						|| dim1.y >= dim2.y + dim2.height
						|| dim1.x + dim1.width <= dim2.x
						|| dim1.x >= dim2.x + dim2.width);

				//Print warning
				if (overlaps)
					System.out.println("Touchable colision warning in " + super.toString() +
							" between component " + i + " and component " + j);
			}
	}
}

File: src/main/kotlin/me/jakejmattson/touchcontrol/utils/TouchableGroup.kt
package me.jakejmattson.touchcontrol.utils

import kotlinx.coroutines.*
import me.jakejmattson.touchcontrol.touchables.Touchable
import org.opencv.core.*
import java.util.Arrays

/**
 * Apply the same function over many Touchable objects.
 *
 * @author JakeJMattson
 */
class TouchableGroup(vararg components: Touchable) {

    private val components: ArrayList<Touchable> = arrayListOf(*components)

    fun addComponent(component: Touchable) = components.add(component)

    fun removeComponent(component: Touchable) = components.remove(component)

    fun getComponents() = components

    fun setColor(color: Scalar) = components.forEach { it.color = color }

    fun updateDetectionPoint(filteredImage: Mat) =
        runBlocking {
            components.forEach {
                launch { it.updateDetectionPoint(filteredImage) }
            }
        }

    fun drawOnto(rawImage: Mat) =
        runBlocking {
            components.forEach {
                launch { it.drawOnto(rawImage) }
            }
        }

    fun performAction() =
        runBlocking {
            components.forEach {
                launch { it.performAction() }
            }
        }

    /**
     * Check whether or not any of the components in
     * the group overlaps (collides) with another.
     *
     * @return Whether or not the group is collision-free
     */
    private fun checkCollision(): Boolean {

        var hasNoCollision = true

        for (i in components.indices)
            for (j in i + 1 until components.size) {

                val dim1 = components[i].dimensions
                val dim2 = components[j].dimensions

                //Check if two components overlap
                val overlaps = !(dim1.y + dim1.height <= dim2.y
                    || dim1.y >= dim2.y + dim2.height
                    || dim1.x + dim1.width <= dim2.x
                    || dim1.x >= dim2.x + dim2.width)

                if (overlaps) {
                    hasNoCollision = false

                    println("Touchable collision warning in ${super.toString()} between component $i and component $j")
                }
            }

        return hasNoCollision
    }

    override fun toString() =
        buildString {
            appendln("Touchable objects in group (${super.toString()}): ${components.size}\n")

            components.forEachIndexed { index, component ->
                appendln("($index)$component")
            }
        }
}